AJ-954 - Add capability to edit a cell in Data table view in Azure

yuliadub opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/AJ-594

This is a draft PR to check if my edits so far look good, I still need to add tests and verify all options but strings when it comes to an edit. 

## Summary of changes:
Adding editing capabilities to WDS data table view, reusing existing GCP components that are already in the code. 

Out of scope for this ticket:
- editing of the primary key entity
- deleting value from column and clearing column (GCP features)
- editing a whole row
- types other then string, number, boolean, json

### Why
- create parity between Azure and GCP workspaces data tables

### Testing strategy
- Work in progress

<!-- ### Visual Aids -->

https://github.com/DataBiosphere/terra-ui/assets/9418602/4d70a12e-3832-475d-acd8-4f33a9b8ea08

json edit: 

https://github.com/DataBiosphere/terra-ui/assets/9418602/c0c7ab31-0990-4179-be50-c7c2f87c07a3

types that are not yet supported: 
![image](https://github.com/DataBiosphere/terra-ui/assets/9418602/b35f661b-bd98-4328-9537-27a94679c5ab)

yuliadub committed:
getting the basics of edit to work

yuliadub committed:
clean up

yuliadub committed:
more clean up

yuliadub committed:
fix build

yuliadub committed:
actually fix build

bethsheets reviewed (changes requested):
I was comparing current GCP UI with WDS datatypes (https://support.terra.bio/hc/en-us/articles/12553682382363-Data-tables-Additional-resources). I think Reference should be updated to Relation, and then we have a few missing datatypes (date, datetime, float). Is there a max number of items for that pop-up modal? Can you fit them in there or should I ask UX for help with expanding? These other datatypes could be a followup ticket if you like.

yuliadub committed:
add a wds specific modal for edit

yuliadub commented:
> I was comparing current GCP UI with WDS datatypes (https://support.terra.bio/hc/en-us/articles/12553682382363-Data-tables-Additional-resources). I think Reference should be updated to Relation, and then we have a few missing datatypes (date, datetime, float). Is there a max number of items for that pop-up modal? Can you fit them in there or should I ask UX for help with expanding? These other datatypes could be a followup ticket if you like.

In Azure, you actually cant change a type of just a single cell, so that option shouldn't be there ( I removed it). That will be reflected in the "change column type" UI which is a different ticket. (but I am sure this context will be really helpful for it so I added a link to this in there)

yuliadub committed:
progress push

yuliadub committed:
merge

yuliadub committed:
continuing to work out the wds edit flow

yuliadub committed:
progress push

yuliadub commented on src/workspace-data/data-table/wds/attribute-utils.js:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  // there is likely a better way to expose this
+  return { error: 'Editing the current type is not supported.' };
```
I could not figure out an easy way to do this in the data table view (i.e. making the edit button not clickable and show a message with "type not supported") open to suggestions on how to do this better. This is the last field I try to edit in my video, it open the modal but it tells you the value cant be edited

bethsheets replied:
Can you remove the edit button completely from that datatype?

yuliadub replied:
@bethsheets  figured out how to do it, would you rather not have it all or keep it disabled with a message?
![image](https://github.com/DataBiosphere/terra-ui/assets/9418602/dc6d325b-bd8e-4861-9e08-94bb32474b76)

bethsheets replied:
I like this! I might update the language to "Editing this data type is not currently supported."

yuliadub commented on src/workspace-data/data-table/wds/attribute-utils.js:
```diff
@@ -0,0 +1,30 @@
+  );
+
+  // if the type matches, editing is allowed
+  if (type === getTypeAttribute.toLowerCase() && dataProvider.features.supportEntityUpdatingTypes.includes(getTypeAttribute.toLowerCase())) {
```
I think I will end up taking this out -- but the idea I had here was to make sure that the data we are getting actually has the correct type (if it doesnt match, then we shoulnd't save). 

currently this follows GCP implementation where it makes null values strings, that wont work for Azure. If the value is null it should always default to type that came from the schema.

yuliadub commented on src/workspace-data/data-table/wds/WDSContent.js:
```diff
@@ -18,19 +17,21 @@ const WDSContent = ({
       entityMetadata,
       googleProject,
-      workspaceId: { namespace, name },
+      workspaceId: { namespace, name, id },
```
I kind of want to rename this, but I am a bit confused how this is different from the "workspace" variable below it

nawatts replied:
Yeah, we can probably get rid of this prop if the entire workspace is already passed to DataTable. DataTable can pull the relevant attributes off of `workspace`.

jladieu replied:
Love this idea.

nawatts commented on src/libs/ajax/WorkspaceDataService.ts:
```diff
@@ -120,6 +121,19 @@ export const WorkspaceData = (signal) => ({
+    instanceId: string,
+    recordType: string,
+    recordId: string,
+    attributes: object
```
This type could be more descriptive than `object`.
```suggestion
    attributes: { [attribute: string]: any }
```

nawatts commented on src/libs/ajax/data-table-providers/WdsDataTableProvider.test.ts:
```diff
@@ -691,6 +707,26 @@ describe('WdsDataTableProvider', () => {
+      .then((actual) => {
+        // Assert
+        expect(actual.attributes).toStrictEqual(expected);
+      });
```
Terra UI mostly uses async/await over Promise callbacks by convention.

```suggestion
    const actual = await provider.updateRecord({ instance: uuid, recordName: 'test', recordId: 'record1', record: expected });

    // Assert
    expect(actual.attributes).toStrictEqual(expected);
```
This suggest also requires marking the test function `async`.

nawatts commented on src/workspace-data/data-table/wds/SingleEntityEditorWds.js:
```diff
@@ -0,0 +1,88 @@
+
+export const SingleEntityEditorWds = ({
+  entityType,
+  entityName,
```
Since this component is specific to WDS, can we avoid carrying over the "entity" naming from Entity Service components?

```suggestion
  recordType,
  recordId,
```

nawatts commented on src/workspace-data/data-table/wds/attribute-utils.test.js:
```diff
@@ -0,0 +1,72 @@
+
+// this test needs to be edited
+describe('getAttributeType', () => {
+  it('returns type of attribute value', () => {
```
This looks like a good place for a parameterized test using `it.each`.
https://jestjs.io/docs/api#testeachtablename-fn-timeout

nawatts reviewed (commented):
It would be nice if new code was TypeScript instead of JavaScript.

jladieu commented on src/libs/ajax/data-table-providers/WdsDataTableProvider.ts:
```diff
@@ -156,7 +163,8 @@ export class WdsDataTableProvider implements DataTableProvider {
       supportsTypeRenaming: false,
       supportsEntityRenaming: false,
-      supportsEntityUpdating: false, // TODO: enable as part of AJ-594
+      supportsEntityUpdating: true,
```
If you rebase onto main (and, with apologies, resolve merge conflicts), you will hopefully get some changes to this file which showcase how I used capabilities backend to dynamically toggle these `supports` flags.  I'm guessing we may want to do something similar for edit attribute.

yuliadub replied:
the way I understand the capabilities bit is: it is possible someone may have a version of WDS that doesnt have a specific API call (like delete or rename), so we need to track this. 

since no backend calls were added to support editing (that API has been there from the start) -- it sounds like this shouldnt be impacted by this. But let me know if I am misunderstanding

yuliadub replied:
chatted more about this with Josh offline - capabilities needed for edit are all already supported, so confirmed that we do no need this for edit. Primarily this is the case because in WDS the edit modal will not be able to support a change of type like it does in GCP - in GCP there is no concept of the whole column being one type (each value in each individual column can end up being a different type, there is no strict typing per column), while in WDS that is required.

jladieu commented on src/workspace-data/data-table/wds/SingleEntityEditorWds.js:
I think this maybe came up earlier as a design decision...so apologies if I'm revisiting, but did you decide that the existing Entity editor logic was divergent enough from what we'll support in WDS that WDS needs its own component?

My instinct would be to attempt to author as little novel component code for this change as possible, and reuse all the existing stuff where provided (providing override seams in the `DataTableProvider` as needed)...but I get it if it's just too different from what WDS can use.

yuliadub replied:
yeah, this is something Nick and I were chatting about earlier. My first stab was to try and use them for both. 

The main reason why I didnt is because the GCP editor lets the user change the type of any column type, which is just not something we support in Azure at all. (and that applies to each value) PLus Azure has different types that we'd need to support too. it seemed like making the UI work for both given that constraint was just going to be too messy.  as I have gotten more familiar with the code, I'll revisit though!

yuliadub replied:
after revisiting sticking with a different modal still seems like the way to go vs trying to make the two work together side by side.

nawatts commented on src/workspace-data/data-table/wds/attribute-utils.js:
```diff
@@ -0,0 +1,30 @@
+import * as Utils from 'src/libs/utils';
+
+export const getAttributeType = (attributeName, entityTypeAttributes, attributeValue, dataProvider) => {
```
I don't think we need the attribute value to get the type for WDS. It should be available from the schema.

yuliadub replied:
see comment https://github.com/DataBiosphere/terra-ui/pull/4588/files#r1472157907 on my thoughts on this 😄

yuliadub committed:
merge

yuliadub committed:
getting there

yuliadub committed:
fixing build

yuliadub committed:
actually fix build

yuliadub committed:
small changes

yuliadub committed:
start of test

yuliadub committed:
Merge branch 'dev' into aj-954-wds-edit-cell

yuliadub commented on src/workspace-data/data-table/wds/WDSContent.test.ts:
```diff
@@ -215,4 +216,21 @@ describe('WDSContent', () => {
+        render(h(WDSContent, props));
+      });
+
+      // todo: need to figure out how to get the right column to check here
```
would appreciate help here - a bit confused how I make it easy to figure out which parts of the table I need to query to see if the icon is disabled or not (per column)

yuliadub committed:
fixed the bug of random undefined column!

yuliadub committed:
fix build

yuliadub committed:
add test

yuliadub committed:
Merge branch 'dev' into aj-954-wds-edit-cell

yuliadub committed:
Merge branch 'dev' into aj-954-wds-edit-cell

yuliadub committed:
fix eslint after merge

yuliadub committed:
edit modal title

yuliadub commented:
per suggestions during today's demo, changing the edit Modal title. fyi @nawatts @bethsheets 
![image](https://github.com/DataBiosphere/terra-ui/assets/9418602/f48386a0-93b2-4238-a984-07b32ebc2c09)

nawatts commented on src/libs/ajax/data-table-providers/DataTableProvider.ts:
```diff
@@ -125,6 +133,7 @@ export interface DataTableFeatures {
   supportsTypeRenaming: boolean;
   supportsEntityRenaming: boolean;
   supportsEntityUpdating: boolean;
+  supportEntityUpdatingTypes?: string[];
```
Super nit: the other fields here are named "supports..." vs "support...".

yuliadub replied:
ah man, thank you for noticing that - if we didnt catch it, it would make so mad :D

nawatts commented on src/libs/ajax/data-table-providers/WdsDataTableProvider.ts:
```diff
@@ -59,13 +60,24 @@ export interface TsvUploadResponse {
+export interface RecordResponseBody {
+  id: string;
+  type: string;
+  attributes: object;
```
Nit. This type could be slightly more descriptive as something like:

```suggestion
  attributes: { [attributeName: string]: any };
```

nawatts commented on src/workspace-data/data-table/shared/AttributeInput.ts:
```diff
@@ -0,0 +1,72 @@
+import { NumberInput, TextInput } from 'src/components/input';
+import * as Utils from 'src/libs/utils';
+
+// can be pulled into shared
```
Comment left over from development?

nawatts commented on src/workspace-data/data-table/shared/AttributeInput.ts:
It's really nice that this could be shared between entity service and WDS tables.

nawatts commented on src/workspace-data/data-table/shared/DataTable.js:
```diff
@@ -612,16 +616,34 @@ const DataTable = (props) => {
+                          })
+                        : editable &&
+                          dataProvider.features.supportsEntityUpdating &&
+                          h(EditDataLinkDisabled, {
```
Does this need a dedicated component? Or can the same thing be accomplished by setting `disabled` and `tooltip` props on EditDataLink?

yuliadub replied:
I wasnt sure if it was okay to override tooltip like that, but yes, that most certainly is much cleaner!

nawatts commented on src/workspace-data/data-table/wds/AttributeInput.ts:
```diff
@@ -0,0 +1,122 @@
+  initialValue: any;
+}
+
+const AttributeInput = ({
+  autoFocus = false,
+  value: attributeValue,
+  attributeName,
+  dataProvider,
+  onChange,
+  recordTypeAttributes,
+}: WDSAttributeInputProps) => {
```
Nit, but when there are a lot of props, I think a style like this makes the component signature easier to see.

```suggestion
const AttributeInput = (props: WDSAttributeInputProps): ReactNode => {
  const {
    autoFocus = false,
    value: attributeValue,
    attributeName,
    dataProvider,
    onChange,
    recordTypeAttributes,
  } = props;
```

nawatts commented on src/workspace-data/data-table/wds/AttributeInput.ts:
```diff
@@ -0,0 +1,122 @@
+import { renderInputForAttributeType } from '../shared/AttributeInput';
+import { getAttributeType } from './attribute-utils';
+
+const defaultValueForAttributeType = (attributeType: string | undefined) => {
```
It would be nice to have to a return type annotated on this function.

```suggestion
const defaultValueForAttributeType = (attributeType: string | undefined): any => {
```

nawatts commented on src/workspace-data/data-table/wds/SingleEntityEditorWds.ts:
```diff
@@ -0,0 +1,105 @@
+  const [isBusy, setIsBusy] = useState(false);
+
+  const modalTitle = `Edit ${attributeName} value for ${recordName}`;
+  const modalDetails = originalValueType ? `Column Type is  ${originalValueType}` : '';
```
Since the title is sentence case, suggest sentence case here as well.

```suggestion
  const modalDetails = originalValueType ? `Column type is ${originalValueType}` : '';
```

nawatts commented on src/workspace-data/data-table/wds/WDSContent.test.ts:
```diff
@@ -213,4 +216,26 @@ describe('WDSContent', () => {
+
+  describe('edit column', () => {
+    it('edit field icon is present for types that support editing', async () => {
+      const { props } = setup({
```
Nit: preferably, all tests have arrange/act/assert sections commented.

```suggestion
      // Arrange
      const { props } = setup({
```

nawatts commented on src/workspace-data/data-table/wds/WDSContent.test.ts:
```diff
@@ -213,4 +216,26 @@ describe('WDSContent', () => {
+        render(h(WDSContent, props));
+      });
+
+      const editableValues = await screen.findAllByText('Edit value');
```
```suggestion
      // Assert
      const editableValues = await screen.findAllByText('Edit value');
```

nawatts commented on src/workspace-data/data-table/wds/attribute-utils.test.ts:
```diff
@@ -0,0 +1,60 @@
+import { getAttributeType } from './attribute-utils';
+
+// this test needs to be edited
```
Is this comment still accurate?

nawatts commented on src/workspace-data/data-table/wds/attribute-utils.ts:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  // there is likely a better way to expose this
+  return { error: 'Editing the current type is not supported.' };
```
This error message seems out of place. This function is for getting the attribute type, not editing.

```suggestion
  throw new Error('Unable to determine attribute type');
```

yuliadub committed:
pr feedback

jladieu commented:
Nice work, this was a biggie'!
